Add Gazelle to generate cue_rules

lenage opened the pull request:
1. support to generate `cue_instance` 
2. only support `github.com/seh/rules_cue/`

Commit:
add cue/deps.bzl file to support noenable_bzlmod

ツ cd examples/bzlmod
ツ bazel build //...
(14:07:29) INFO: Invocation ID: f93ddc7e-a953-4b97-9c2b-093b8277a626
(14:07:29) INFO: Current date is 2025-01-21
(14:07:33) INFO: Analyzed 4 targets (0 packages loaded, 0 targets configured).
(14:07:33) INFO: Found 4 targets...
(14:07:33) INFO: Elapsed time: 3.547s, Critical Path: 0.01s
(14:07:33) INFO: 1 process: 1 internal.
(14:07:33) INFO: Build completed successfully, 1 total action

Commit:
make build pass for old workspace

Commit:
move cue_test to cue.bzl

lenage committed:
Merge branch 'seh:main' into main

Commit:
gazelle: load cue_instance and other com_github_seh_cue_rules's rules

lenage committed:
Update cue/deps.bzl

Co-authored-by: Yujun Zhang <zhangyujun@gmail.com>

yujunz commented on gazelle/cue/resolve.go:
```diff
@@ -0,0 +1,138 @@
+	"tool/file":       true,
+	"tool/http":       true,
+	"tool/os":         true,
+}
```
What is the source of truth for this list? Is it manually maintained or generated from corresponding cue version?

lenage replied:
manually maintained  from cue stdlib

Commit:
update cue.go and generate.go to support new rules

Commit:
update generate.go and testdata

Commit:
generate BUILD file for sub folder

Commit:
add cue_enable_tnarg_rules_cue to control tnarg_rules_cue generation

Commit:
update testdata BUILD file after add cue_enable_tnarg_rules_cue flag

Commit:
generate correct BUILD rule for cue.mod

Commit:
Refactor findNearestCueModule to simplify directory traversal logic

The function now uses a more concise approach to search for a cue.mod directory, reducing complexity while maintaining the same functionality of finding the nearest cue module.

Commit:
fix: handle generate instance name

Commit:
update cue_exported_files rule

Commit:
update cue.mod resolve

Commit:
update test gazelle BUILD file

Commit:
update BUILD file

Commit:
Update Bazel visibility and add cue_exported_files rule

- Modify cue_instance visibility in main BUILD file to use subpackages
- Add cue_exported_files rule for dev environment with module and dependency configuration
- Update load statement to include new cue_exported_files rule

Commit:
Add CUE output format configuration and enhance rule generation

- Introduce `cue_output_format` configuration option with default set to "json"
- Update `cueConfig` to include output format field
- Modify rule generation to support output format in exported instances and files
- Refactor `GenerateRules` method to use a context-based approach for better modularity
- Add support for generating consolidated instances with CUE output format
- Update visibility of generated rules to use public visibility
- Enhance rule generation to handle various CUE file scenarios more comprehensively

Commit:
Update CUE module exported files with consistent configuration

- Add `cue_exported_files` rule to multiple CUE module directories
- Set consistent output format to JSON for exported files
- Add module and dependency references for generated rules
- Enhance visibility configuration for generated rules
- Standardize rule generation across different CUE module packages

Commit:
Enhance CUE Gazelle plugin with improved exported files and module resolution

- Add `srcs` attribute support for `cue_exported_files` rule
- Modify exported files and instances generation to include source files
- Improve module resolution logic for different rule kinds
- Update naming convention for exported files and instances
- Enable debug logging for module indexing
- Add comprehensive documentation for `cueExportedFiles` struct

Commit:
Improve CUE import path resolution with prefix support

- Add prefix support for CUE module imports in Gazelle plugin
- Modify import path generation to use optional module prefix
- Disable debug module indexing by default
- Update import resolution for cue_instance rules to handle prefixed imports

Commit:
Update CUE exported files and instances across multiple modules

- Standardize naming convention for `cue_exported_files` and `cue_exported_instance` rules
- Add `srcs` attribute to specify source files for exported rules
- Update module references and visibility configurations
- Ensure consistent JSON output format for exported files
- Modify rule generation across various CUE module packages and API directories

Commit:
remove duplicated exported_instance rule

Commit:
Enhance CUE module indexing and import resolution

- Refactor module indexing to support multiple subdirectories (gen, usr, pkg)
- Improve import path resolution across different CUE module directories
- Update cueModIndex to use CueModuleInfo as key for more flexible indexing
- Add more robust handling of import path matching and resolution
- Modify debug logging for module indexing
- Simplify and optimize import resolution logic

23:23:14 rules_cue ツ bazel build test/testdata/gazelle/...
(23:23:19) INFO: Invocation ID: a286d67c-b67a-4e4f-b2b7-101830bf849a
(23:23:19) INFO: Current date is 2025-03-03
(23:23:19) INFO: Analyzed 297 targets (62 packages loaded, 716 targets configured).
(23:23:19) INFO: Found 297 targets...
(23:23:19) INFO: Elapsed time: 0.212s, Critical Path: 0.02s
(23:23:19) INFO: 1 process: 1 internal.
(23:23:19) INFO: Build completed successfully, 1 total action

Commit:
Remove cue_exported_instance from BUILD files

- Remove redundant `cue_exported_instance` load statements across multiple CUE module directories
- Simplify Bazel load statements to use only `cue_exported_files` and `cue_instance`
- Update source file configurations for exported files
- Maintain consistent JSON output format and visibility settings

Commit:
Refine CUE Gazelle plugin dependency resolution and configuration

- Disable `cue_enable_tnarg_rules_cue` by default in configuration
- Update dependency resolution to use instance labels for specific rule kinds
- Simplify label generation for CUE module dependencies
- Remove redundant library label generation logic

Commit:
update generator to not add exported_files by default

- Remove `load("@rules_cue//cue:cue.bzl", "cue_exported_files", "cue_instance")` from multiple BUILD files
- Simplify Bazel load statements across various CUE module directories
- Maintain consistent rule generation and configuration
- Reduce redundancy in Bazel build configuration files

Commit:
add fix to handle cue_library and update name coversion

Commit:
do not remove cue_library in fix

Commit:
Remove aliases.bzl load statement from Gazelle CUE plugin

- Remove unnecessary load statement for @rules_cue//cue:aliases.bzl
- Simplify Gazelle CUE plugin configuration
- Clean up unused import references

Commit:
Enhance cue_library import handling in Gazelle CUE plugin

- Add conditional indexing for cue_library rules based on configuration setting `enableTnargRulesCue`
- Ensure import path is validated before indexing
- Improve debug logging for import resolution to include import paths

Commit:
Refactor CUE file parsing and indexing in Gazelle CUE plugin

    - Replace manual package name extraction with CUE parser for improved accuracy
    - Introduce new functions for indexing import paths, including colon-style and domain-specific imports
    - Enhance debug logging for better traceability during import resolution
    - Streamline import specification creation to support colon-style imports

Commit:
remove cue_library to use new cue_instance rule

Commit:
add fixes to remove cue_library

Commit:
craft package import path

Commit:
Refactor CUE rule definitions and enhance Gazelle CUE plugin configuration

- Introduce aliases for cue_consolidated_instance and cue_exported_instance for improved usability
- Update cue_register_toolchains function signature for clarity
- Enable `cue_enable_tnarg_rules_cue` by default in configuration
- Adjust rule generation logic to streamline empty rule creation

Commit:
Update CUE rule definitions to include cue_consolidated_instance

- Add cue_consolidated_instance rule to various BUILD files for improved CUE instance handling
- Modify existing cue_instance rules to support consolidated definitions
- Enhance .gitignore to exclude log files
- Refactor cue.bzl to streamline relative instance path handling

Commit:
Add PrometheusRule CUE definitions and BUILD configuration

- Introduce new CUE definitions for the PrometheusRule custom resource in types_gen.cue
- Create corresponding BUILD.bazel file to define cue_instance and cue_consolidated_instance for the PrometheusRule
- Ensure visibility is set to public for both instances

Commit:
Add time package CUE definitions and BUILD configuration

- Introduce new CUE definitions for the time package, including time_go_gen.cue, format_go_gen.cue, zoneinfo_go_gen.cue, and others.
- Create corresponding BUILD.bazel file to define cue_instance and cue_consolidated_instance for the time package.
- Implement a workaround in hack.cue for a known issue in the CUE language.
- Ensure visibility is set to public for all instances.

Commit:
[cue] add cue_cmd to easy run cue {fmt,cmd} etc...

- Add cmd.sh.tpl to the BUILD.bazel file for better script management.
- Update cue-run-from-runfiles to improve error handling and command execution based on the subcommand.
- Modify cue.bzl to include a commented-out embed field for potential future use.
- Introduce cue_cmd in BUILD files for hello_world and gazelle test data to streamline command execution and visibility settings.

Commit:
update test buildfile

Commit:
Deprecate 'command' attribute in cue_cmd rule in favor of 'tool' for improved clarity and backward compatibility. Update comments to reflect changes.

Commit:
update generateEmpty to remove old target

Commit:
add rel and pth info when parseCUE files

Commit:
add cue_test_golden_suffix directive

Commit:
[test] add cue_test requires data prepare

Commit:
Merge branch 'main' into gazelle

Commit:
[feat] genrate cue_test with golden_filename directive

Commit:
cleanup debug info

Commit:
Ensure unique exported instance creation in cue generation

Commit:
Remove appending of exported instance rules in cue generation

Commit:
rename cue_test target name

Commit:
handle cue_instance ancestor for cue_instance

Commit:
fix ancestor path issue

Commit:
[runfile] add CUE_DEBUG=sortfields by default

Commit:
update ancestor path in BUILD file and adjust file permissions for golden JSON.

Commit:
[fix] recursively lookup to find cue ancestors

Commit:
[fix] cleanup logs

lenage commented:
merge to main manually due to conflict with upstream, this PR closed
